pkg/gardenlet/controller/shoot/shoot_control_prepare_migration.go

```diff
@@ -0,0 +1,298 @@
+			Fn:   flow.TaskFn(botanist.DeploySecrets),
+		})
+		deployBackupEntryInGarden = g.Add(flow.Task{
+			Name: "Deploying backup entry",
+			Fn:   flow.TaskFn(botanist.DeployBackupEntryInGarden),
+		})
+		wailtUntilBackupEntryInGardenReconciled = g.Add(flow.Task{
+			Name:         "Waiting until the backup entry has been reconciled",
+			Fn:           flow.TaskFn(botanist.WaitUntilBackupEntryInGardenReconciled),
+			Dependencies: flow.NewTaskIDs(deployBackupEntryInGarden),
+		})
+		deployETCD = g.Add(flow.Task{
+			Name:         "Deploying main and events etcd",
+			Fn:           flow.TaskFn(botanist.DeployETCD).RetryUntilTimeout(defaultInterval, defaultTimeout),
+			Dependencies: flow.NewTaskIDs(deploySecrets, wailtUntilBackupEntryInGardenReconciled),
+		})
+		waitUntilEtcdReady = g.Add(flow.Task{
+			Name:         "Waiting until main and event etcd report readiness",
+			Fn:           flow.TaskFn(botanist.WaitUntilEtcdReady),
+			Dependencies: flow.NewTaskIDs(deployETCD),
+		})
+		prepareControlPlaneDeployments = g.Add(flow.Task{
+			Name:         "Stop Kubernetes API Server in Shoot's namespace",
+			Fn:           flow.TaskFn(botanist.PrepareControlPlaneDeployments).RetryUntilTimeout(defaultInterval, defaultTimeout).SkipIf(shootHibernated),
+			Dependencies: flow.NewTaskIDs(waitUntilEtcdReady),
+		})
+		waitUntilAPIServerIsDown = g.Add(flow.Task{
+			Name:         "Wait for Kubernetes API Server to be scaled to zero",
+			Fn:           flow.TaskFn(botanist.WaitUntilKubeAPIServerScalesDown).SkipIf(shootHibernated),
+			Dependencies: flow.NewTaskIDs(prepareControlPlaneDeployments),
+		})
+		_ = g.Add(flow.Task{
+			Name:         "Destroying ingress DNS record",
+			Fn:           flow.TaskFn(botanist.DestroyIngressDNSRecord).DoIf(dnsEnabled),
+			Dependencies: flow.NewTaskIDs(waitUntilAPIServerIsDown),
+		})
+		_ = g.Add(flow.Task{
+			Name:         "Destroying external domain DNS record",
+			Fn:           flow.TaskFn(botanist.DestroyExternalDomainDNSRecord).DoIf(dnsEnabled),
+			Dependencies: flow.NewTaskIDs(waitUntilAPIServerIsDown),
+		})
+		_ = g.Add(flow.Task{
+			Name:         "Destroying internal domain DNS record",
+			Fn:           flow.TaskFn(botanist.DestroyInternalDomainDNSRecord).DoIf(dnsEnabled),
+			Dependencies: flow.NewTaskIDs(waitUntilAPIServerIsDown),
+		})
+		createETCDSnapshot = g.Add(flow.Task{
+			Name:         "Create ETCD Snapshot",
+			Fn:           botanist.CreateETCDSnapshot,
+			Dependencies: flow.NewTaskIDs(prepareControlPlaneDeployments, waitUntilAPIServerIsDown),
+		})
+		_ = g.Add(flow.Task{
+			Name:         "Scale ETCD to zero",
+			Fn:           botanist.ScaleETCDStatefulSetsToZero,
+			Dependencies: flow.NewTaskIDs(createETCDSnapshot),
+		})
+		_ = g.Add(flow.Task{
+			Name:         "Annotate Extensions with operation - migration",
+			Fn:           botanist.AnnotateExtensionCRDsForMigration,
+			Dependencies: flow.NewTaskIDs(waitUntilAPIServerIsDown),
+		})
+		_ = g.Add(flow.Task{
```

Comment:
Shouldn't the namespace be cleaned up in the source seed? How will this happen?

Reply:
As discussed with @mandelsoft yesterday and also proposed by you, @swilen-iwanow, the desired option is to have a separate "garbage collection/cleanup" controller that scans the namespaces in the seed for shoot namespaces that belong to shoots that are no longer scheduled on this seed.
You should be able to cleanup the namespace by running the shoot deletion flow (everything that could cause harm (e.g., infrastructure extension should not really be deleted) should already be gone after annotating the extension CRDs for migration in the prepare migration flow). Also, it probably makes sense to only start with the namespace cleanup if the `Shoot`'s `.status.lastOperation.type` is no longer `Restore`, i.e., after the restoration in the new seed has completed successfully. WDYT?

Reply:
Yes sounds good, I think a general solution for cases with orphaned resources will be also good if you start migrating to the target Seed that will hit some limits and you have to change the target again. (hint from Nachev)